modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureReader.java

```diff
@@ -0,0 +1,218 @@
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.geotools.data.DataStore;
+import org.geotools.data.Query;
+import org.geotools.data.simple.SimpleFeatureCollection;
+import org.geotools.data.simple.SimpleFeatureIterator;
+import org.geotools.data.simple.SimpleFeatureReader;
+import org.geotools.feature.FeatureIterator;
+import org.geotools.feature.simple.SimpleFeatureBuilder;
+import org.geotools.util.logging.Logging;
+import org.opengis.feature.Feature;
+import org.opengis.feature.simple.SimpleFeature;
+import org.opengis.feature.simple.SimpleFeatureType;
+import org.opengis.filter.Filter;
+
+/** {@link DataStore} for a vector mosaic. */
+public class VectorMosaicFeatureReader implements SimpleFeatureReader {
+    static final Logger LOGGER = Logging.getLogger(VectorMosaicFeatureReader.class);
+    private final SimpleFeatureCollection delegateCollection;
+    private final Query query;
+    private final VectorMosaicFeatureSource source;
+    private final SimpleFeatureType granuleFeatureType;
+    FeatureIterator delegateIterator;
+    FeatureIterator granuleIterator;
+    DataStore granuleDataStore;
+    private Feature nextGranule = null;
+    SimpleFeature delegateFeature = null;
+
+    /**
+     * Constructor
+     *
+     * @param delegateFeatures index features from the delegate store
+     * @param query query to use to filter the index features and the granule features
+     * @param vectorMosaicFeatureSource the source of the vector mosaic
+     */
+    public VectorMosaicFeatureReader(
+            SimpleFeatureCollection delegateFeatures,
+            Query query,
+            VectorMosaicFeatureSource vectorMosaicFeatureSource) {
+        this.delegateCollection = delegateFeatures;
+        this.query = query;
+        this.source = vectorMosaicFeatureSource;
+        delegateIterator = delegateCollection.features();
+        granuleFeatureType = getGranuleType(delegateCollection);
+    }
+
+    /**
+     * Get the granule feature type from the delegate collection by pulling one granule feature
+     *
+     * @param delegateCollection the delegate collection
+     * @return the granule feature type
+     */
+    private SimpleFeatureType getGranuleType(SimpleFeatureCollection delegateCollection) {
+        try (SimpleFeatureIterator fi = delegateCollection.features(); ) {
+            SimpleFeature firstDelegateFeature = null;
+            if (fi.hasNext()) {
+                firstDelegateFeature = fi.next();
+            }
+            if (firstDelegateFeature == null) {
+                throw new IllegalArgumentException("No delegate features found");
+            }
+            VectorMosaicGranule granule =
+                    VectorMosaicGranule.fromDelegateFeature(firstDelegateFeature);
+            source.initGranule(granule, true);
+            return granule.getDataStore().getSchema(granule.getGranuleTypeName());
```

Comment:
Here a datastore is created but not disposed of: making it closeable should make sure that you don't lose bits along the road.

Reply:
Fixed